src/sinks/clickhouse/arrow_schema.rs

```diff
@@ -0,0 +1,767 @@
+fn parse_schema_from_response(response: &str) -> crate::Result<Arc<Schema>> {
+    let mut columns: Vec<ColumnInfo> = Vec::new();
+
+    for line in response.lines() {
+        if line.trim().is_empty() {
+            continue;
+        }
+
+        let column: ColumnInfo = serde_json::from_str(line)
+            .map_err(|e| format!("Failed to parse column info: {}", e))?;
+        columns.push(column);
+    }
+
+    if columns.is_empty() {
+        return Err("No columns found in table schema".into());
+    }
+
+    let mut fields = Vec::new();
+    for column in columns {
+        let (arrow_type, nullable) = clickhouse_type_to_arrow(&column.column_type)
+            .map_err(|e| format!("Failed to convert column '{}': {}", column.name, e))?;
+        fields.push(Field::new(&column.name, arrow_type, nullable));
+    }
+
+    Ok(Arc::new(Schema::new(fields)))
+}
+
+/// Unwraps ClickHouse type modifiers like Nullable() and LowCardinality().
+/// Returns a tuple of (base_type, is_nullable).
+/// For example: "Nullable(LowCardinality(String))" -> ("String", true)
+fn unwrap_type_modifiers(ch_type: &str) -> (&str, bool) {
+    let mut base = ch_type;
+    let mut is_nullable = false;
+
+    // Check for Nullable wrapper
+    if let Some(inner) = base.strip_prefix("Nullable(") {
+        is_nullable = true;
+        base = inner.strip_suffix(')').unwrap_or(inner);
+    }
+
+    // Check for LowCardinality wrapper
+    if let Some(inner) = base.strip_prefix("LowCardinality(") {
+        base = inner.strip_suffix(')').unwrap_or(inner);
+    }
+
+    (base, is_nullable)
+}
+
+fn unsupported(ch_type: &str, kind: &str) -> String {
+    format!(
+        "{kind} type '{ch_type}' is not supported. \
+         ClickHouse {kind} types cannot be automatically converted to Arrow format."
+    )
+}
+
+fn clickhouse_type_to_arrow(ch_type: &str) -> Result<(DataType, bool), String> {
+    let (base_type, is_nullable) = unwrap_type_modifiers(ch_type);
+
+    let data_type = match base_type {
+        // String types
+        "String" => DataType::Utf8,
+        _ if base_type.starts_with("FixedString") => DataType::Utf8,
+
+        // Integer types
+        "Int8" => DataType::Int8,
+        "Int16" => DataType::Int16,
+        "Int32" => DataType::Int32,
+        "Int64" => DataType::Int64,
+        "UInt8" => DataType::UInt8,
+        "UInt16" => DataType::UInt16,
+        "UInt32" => DataType::UInt32,
+        "UInt64" => DataType::UInt64,
+
+        // Floating point types
+        "Float32" => DataType::Float32,
+        "Float64" => DataType::Float64,
+
+        // Boolean
+        "Bool" => DataType::Boolean,
+
+        // Date and time types (timezones not currently handled, defaults to UTC)
+        "Date" | "Date32" => DataType::Date32,
+        "DateTime" => DataType::Timestamp(TimeUnit::Second, None),
+        _ if base_type.starts_with("DateTime64") => parse_datetime64_precision(base_type)?,
```

Comment:
Using `starts_with` here is prone to someone using `DateTime64FooBarBaz` and getting back a DateTime64. Again I'm worried about two things here: first is untrusted input, but not sure if applicable; second is that if some other `DateTime64` type is added like `DateTime64UTC` or something like that then this becomes a real problem. You already implemented `extract_identifier`, so the simplest solution is to move all these to a `_ => {` block and matching on `extract_identifier().0` instead of using `str::starts_with`.

This is of course relevant in the other matches that are using `starts_with`